readme.md
FTP-Account to UserMeta
=======================

This WordPress-Plugin stores the ftp-account data of a single user in the usermeta. This prevent to insert the ftp-account data everytime WordPress need them.
It also make it possible to use more than one ftp-user on the same WordPress-Installation.

Kudos
-----
Some parts of this code are originally by Frank Bueltge [www.bueltge.de]

Inspiration by discussion from Sergej Mueller [http://playground.ebiene.de]

Feedback from:
Franz-Josef Kaiser [https://github.com/franz-josef-kaiser]

changelog.txt
CHANGELOG:

23.05.2012
	v0.1.2
		- remove methods loadtextdomain and get:plugin_data for more simplicity
		- using current_user_can instead of $user->has_cap() to decide if the plugin should start or not
		- move add_filter( 'request_filesystem_credentials' ) to method add_admin_hooks()

20.05.2012
	v0.1.1	- Change cleanup after plugin uninstall
	
19.05.2012
	v0.1	- First release.


classes/class-formatter.php
<?php
/**
 * Simple Template Engine
 *
 * PHP version 5.3
 *
 * @category   PHP
 * @package    Template_Engine
 * @author     Ralf Albert <me@neun12.de>
 * @license    GPLv3 http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.1.1
 * @link       http://neun12.de
 */

/**
 * 
 * Formtatter
 *  
 * Formatter is a class which provide a simple template engine. It uses the printf/sprintf-syntax to format
 * values.
 * 
 * Usage:
 * $values = array(
 * 	'key_1'	=> 'value_1',
 * 	'key_2'	=> 'value_2',
 * );
 * 
 * $format = 'key_1 have value %key_1%. key_2 have value %key_2%. %key_1% and %key_2%';
 * Formatter::printf( $format, $values );
 * 
 * $values = new stdClass();
 * $values->key_1 = 'value 1';
 * $values->key_2 = 'value 2';
 * $format = '>key_1< belongs to key_1, >key_2< belongs to key_2';
 * 
 * Formatter::set_delimiter( '<', '>' );
 * $e = Formatter::sprintf( $format, $values );
 * 
 * $values = array(
 * 	'number_one'	=> 3,
 *  'number_two'	=> 5,
 *  );
 *  
 * $format	 = 'Number One filled to 5 positions: %number_one[05d]%';
 * $format	.= ' and Number Two as hex (upper chars) filled to 6 positions: %number_two[06X]%';
 * Formatter::printf( $format, $values );
 *  
 *    
 * Formatter::printf( string $format, array|object $values );
 * Formatter::sprintf( string $format, array|object $values );
 * Formatter::set_delimiter( string $start_delimiter, string $end_delimiter );
 * 
 * @author Ralf Albert
 * @version 1.1.1
 * @see http://php.net/manual/function.sprintf.php
 */
 
class Formatter
{
	/**
	 * 
	 * Starting delimiter
	 * @var string
	 */
	public static $start_delimiter	= '%';
	
	/**
	 * 
	 * Ending delimiter
	 * @var string
	 */
	public static $end_delimiter	= '%';
	
	/**
	 * 
	 * Replacing values in a format-string
	 * @param string $format
	 * @param array|object $values
	 * @throws Exception
	 * @return string|bool	Returns the formated string or FALSE on failure
	 */
	public static function sprintf( $format = '', $values = NULL ){
		/*
		 * Checking arguments
		 */
		if( empty( $format ) || NULL == $values )
			return FALSE;
			
		if( ! is_string( $format ) )
			throw new Exception( 'Format must be a string' );
			
		if( ! is_array( $values ) && ! is_object( $values ) )
			throw new Exception( 'Values must be type of array or object' );
			
		/*
		 * Do the replacement
		 */	
		foreach( $values as $key => $value ){

			$matches	= array();
			$search_key	= sprintf( '%s%s%s', self::$start_delimiter, $key, self::$end_delimiter );
			$pattern	= sprintf( '/%%%s\[(.*)\]%%/iU', $key );

			// search for the values in format-string. find %key% or %key[format]%
			preg_match_all( $pattern, $format, $matches );
			
			// the '[format]' part was not found. replace only the key with the value 
			if( empty( $matches[1] ) ){
				$format = str_replace( $search_key, $value, $format );
			}
			// one or more keys with a '[format]' part was found.
			// walk over the formats and replace the key with a formated value
			else {

				foreach( $matches[1] as $match ){
					$replace = sprintf( '%' . $match, $value );
					$search = sprintf( '%s%s[%s]%s', self::$start_delimiter, $key, $match, self::$end_delimiter );
					$format = str_replace( $search, $replace, $format );
				}
			}
		}

		// return the formatted string	
		return $format;
		
	}
	
	/**
	 * 
	 * Print a formated string
	 * @param string $format
	 * @param array|object $values
	 * @uses Formatter::sprintf()
	 * @return void
	 */
	public static function printf( $format, $values ){
		echo self::sprintf( $format, $values );
	}
	
	/**
	 * 
	 * Set the start- and end-delimiter
	 * @param string $start
	 * @param string $end
	 */
	public static function set_delimiter( $start = '%', $end = '%' ){
		self::$start_delimiter	= $start;
		self::$end_delimiter	= $end;
	}
}

classes/class-wp_autoloader.php
<?php
/**
 * 
 * WP-Autoloader
 * 
 * Easy to use autoloader
 * 
 * To overwrite $class_prefix, an empty string have to be passed to the autoloader
 * 
 * @author Ralf Albert (neun12@googlemail.com)
 * @version 0.3
 */

/*
 * Usage
 * =====
 * 
 * $autoloader = new WP_Autoloader( [$config] );
 * 
 * $config is optional
 * $config could be array or object
 * 
 * $config should look like this:
 * 
 * $config = new stdClass();
 * $config->abspath = __FILE__;
 * $config->include_path = array( '/lib' );
 * $config->class_extension = '-class.php';
 * $config->class_prefix = array();
 *  
 *  OR
 *  
 * $config = array(  
 * 		'abspath'			=> __FILE__,
 * 		'include_path' 		=> array( 'models', 'views' ),
 * 		'class_extension'	=> '.php',
 * 		'class_prefix' 		=> array( 'model-', 'view-' ),
 * );
 * 
 */

class WP_Autoloader
{
	/**
	 * 
	 * Absolute path to include directory
	 * @var string
	 */
	public static $abspath = FALSE;
	
	/**
	 * 
	 * Relative path to include directory
	 * @var array
	 */
	public static $include_pathes	= array();
	
	/**
	 * 
	 * Extension for class files
	 * @var string
	 */
	public static $class_extension	= '.php';
	
	/**
	 * 
	 * Prefix for class files
	 * @var array
	 */
	public static $class_prefixes 	= array();

	/**
	 * 
	 * Constructor for one-step-autoloading
	 * 
	 * @param array|object $config
	 */
	public function __construct( $config = NULL ){
		if( NULL === $config )
			$config = array();
						
		// convert $config to an array if an object was given and merge defaults into the config-array
		$config = (array) $config;

		$defaults = array(
			'include_path' 		=> self::$include_pathes,
			'class_extension' 	=> self::$class_extension,
			'class_prefix' 		=> self::$class_prefixes 
		);

		extract( array_merge( $defaults, $config ) );
		
		if( isset( $abspath ) && is_string( $abspath ) );
			self::set_abspath( $abspath );
		
		self::autoloader_init( $include_path, $class_prefix, $class_extension );
	}
	
	/**
	 * 
	 * Initialize the spl_autoloader
	 * Overwrite class-vars, set&sanitize include-path, checks if include-path was already set
	 * 
	 * @param array $include_path
	 * @param array $class_prefix
	 * @param string $class_extension
	 */
	public static function autoloader_init( $include_pathes = array(), $class_prefixes = array(), $class_extension = FALSE ){

		if( FALSE === self::$abspath )
			self::$abspath = dirname( __FILE__ );
		
		if( ! is_array( $include_pathes ) )
			$include_pathes = (array) $include_pathes;
			
		if( ! empty( $include_pathes ) )
			self::set_includepath( $include_pathes );
			
		if( ! is_array( $class_prefixes ) )
			$class_prefixes = (array) $class_prefixes;
			
		if( ! empty( $class_prefixes ) )
			self::$class_prefixes = $class_prefixes;

		if( FALSE !== $class_extension )
			self::$class_extension = $class_extension;
  		
		/*
		 * From php.net/spl_autoload (http://de3.php.net/manual/de/function.spl-autoload.php)
		 * 
		 * 1. Add your class dir to include path
		 * 2. You can use this trick to make autoloader look for commonly used "My-class.php" type filenames
		 * 3. Use default autoload implementation or self defined autoloader 
		 */

		foreach( self::$include_pathes as $includepath ){

			$path = self::$abspath . DIRECTORY_SEPARATOR . $includepath;

			// check if the path have already been added to include_path
			$pathes = explode( PATH_SEPARATOR, get_include_path() );

			if( ! in_array( $path, $pathes ) )				
				// set our path at the first position. require, include, __autoload etc. start searching in the first path
				// with our custom path at the first, PHP does not have to search in all other pathes for our classes
				set_include_path( $path . PATH_SEPARATOR . get_include_path() );
			
		}
			
		spl_autoload_extensions( self::$class_extension );
		spl_autoload_register( array( __CLASS__, 'autoload' ) );

	}
		
	/**
	 * 
	 * Callback for spl_autoload_register
	 * 
	 * @param string $class_name
	 */
	private static function autoload( $class_name ){
		// if a class-prefix is set, add it to the class-name
		$load_error = FALSE;
		
		if( ! empty( self::$class_prefixes ) ){
			foreach( self::$class_prefixes as $prefix ){
				$test_class_name = $prefix . $class_name;
				
				try {
					spl_autoload( $test_class_name );
				} catch ( Exception $e ) {
					$load_error = TRUE;
				}
						
			}
			
			if( $load_error )
				throw new Exception( 'Class ' . $class_name . ' not found' );
				
		} else {
				try {
					spl_autoload( $class_name );
				} catch ( Exception $e ) {
					throw new Exception( 'Class ' . $class_name . ' not found' );
				}
		}
		
	}
	
	/**
	 * 
	 * Set the absolute path to file
	 * 
	 * @param string $abspath
	 * @return bool True on success (abspath is an file or directory), false on fail
	 */
	public static function set_abspath( $abspath = FALSE ){
		if( FALSE != $abspath && is_string( $abspath ) ){
			
			if( is_file( $abspath ) ){	
				self::$abspath = dirname( $abspath );
			}
			elseif( is_dir( $abspath) ){
				self::$abspath = $abspath;
			}
			else
				return FALSE;
		}
		
		return TRUE;  
	}
	
	/**
	 * 
	 * Sanitize and set the include path variable
	 * Returns the include path if it is a directory. Returns false if failed
	 * 
	 * @param array $pathes
	 * @return bool True if $path is a file or directory (siccess). False if not (fail)
	 */
	public static function set_includepath( $pathes = array() ){
		if( empty( $pathes ) )
			return FALSE;

		$sanitized_pathes = array();
		
		foreach( $pathes as $path ){
			if( is_string( $path ) ){
				// strip slashes and backslashes at the start and end of the string /classes/ -> classes; /lib/classes/ -> lib/classes
				$path = preg_replace( "#^[/|\\\]|[/|\\\]$#", '', $path );
				
				// replace slashes and backslashes with the OS specific directory seperator
				$path = preg_replace( "#[/|\\\]#", DIRECTORY_SEPARATOR, $path );
				
				// check if the directory exists
				if( ! is_dir( self::$abspath . '/' . $path ) )
					continue;
	
				// $path is a file or directory and was successfully converted
				array_push( $sanitized_pathes, $path );
			}	
		}

		self::$include_pathes = $sanitized_pathes;
		
		unset( $sanitized_pathes );
	}

}


classes/class-wp_encryption.php
<?php
/**
 * Class to encrypt and decrypt in WordPress
 *
 * PHP version 5.3
 *
 * @category   PHP
 * @package    WordPress
 * @subpackage WP Encrypt
 * @author     Ralf Albert <me@neun12.de>
 * @license    GPLv3 http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.1
 * @link       http://wordpress.com
 */

class WP_Encrypt
{
	
	/**
	 * 
	 * Key to encrypt and decrypt data 
	 * @var string
	 */
	private $key = '';
	
	/**
	 * 
	 * The method to use for encryption/decryption
	 * @var string
	 */
	private $method = '';
	
	/**
	 * 
	 * Available methods to encrypt/decrypt
	 * @var array
	 */
	private $available_methods = array(
									'mysql'
								);
								
	/**
	 * 
	 * Default method for encryption/decryption
	 * @var string
	 */
	private $default_method = 'mysql';
	
	public function construct( $method = 'mysql' ){
		
		// sanitize the method to use for cryption
		if( '' != $method ){
			
			$method = strtolower( $method );
			
			if( in_array( $method, $this->available_methods ) )
				$this->method = $method;
			else
				$this->method = $this->default_method;
			
		} else {
			
			$this->method = $this->default_method;
			
		}
		
	}
	
	/**
	 * 
	 * Returns an array with available cryption-methods
	 * @return array $available_methods
	 * @access public
	 * @since 0.1
	 */
	public function get_methods(){
		
		return $this->available_methods;
		
	}
	
	public function set_key( $key = '' ){
		
		if( ! is_string( $key ) )
			return FALSE;
		else
			$this->key = $key;
			
	}
	
	public function encrypt( $string ){
		
	}
	
	public function decrypt( $string ){
		
	}
	
	
}

classes/class-wp_simple_html.php
<?php
/**
 * HTML-Elements for WordPress
 *
 * PHP version 5.3
 *
 * @category   PHP
 * @package    Template_Engine
 * @subpackage WP_Simple_HTML
 * @author     Ralf Albert <me@neun12.de>
 * @license    GPLv3 http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.1
 * @link       http://wordpress.com
 */

/**
 * 
 * Abstract class WP Simple Templater creates a copy of the template-class and
 * provide the templates to the template-engine
 * @author Ralf Albert
 *
 */
abstract class WP_Simple_Templater extends Formatter
{
	/**
	 * 
	 * Instance of template-class
	 * @var object $templates_object
	 */
	protected $templates_object = NULL;
	
	/**
	 * 
	 * Constructor
	 * Creates an instance of the template-class and setup the delimiters for Formatter
	 * @param WP_Simple_Templates $templates
	 */
	public function __construct( WP_Simple_Templates $templates ){
		
		$this->templates_object = &$templates;
		
		$this->set_delimiter( '%', '%' );
		
	}
	
	/**
	 * 
	 * Return a template by given type. Returns an wp-error on failure
	 * @param string $type
	 * @return string|bool Return a template-string if it was defined by the template-class. Or false if no such template was defined
	 */
	protected function get_templates( $templatetype = '', $type = '' ){
		
		// error message
		$err_msg = '';
		
		// if the sub-type is empty (e.g the template is a string not an array), use the template-type as sub-type
		if( '' == $type )
			$type = $templatetype;
		
		// first check if a template-type was set
		if( '' == $templatetype )
			$err_msg = 'Empty template-type';
		
		// check if a template-class was loaded
		elseif( NULL === $this->templates_object )
			$err_msg = 'No templates defined in <strong>' . get_class( $this ) . '</strong>';					

		// we have a template-type and a template-class. get the template(s)
		else {
			
			 $template = $this->templates_object->get_templates( $templatetype );

			 // get error-message thrown by template-class
			 if( is_wp_error( $template ) )
			 	$err_msg = $template->get_error_message();

			 // check the template.
			 // convert a string into an array. else check if the requested type is in the template-array
			 else {
			 	
				 if( ! is_array( $template ) )
				 	$template[$type] = $template;
	
				 elseif( ! in_array( $type, array_keys( $template ) ) )
				 	$err_msg = $type . ' is not defined.';

			 }
			 
		}

		// if a error occurs, return an error-object
		if( '' != $err_msg ){
			
			return new WP_Error(
					'template_error',
					sprintf(
						'<div class="error"><h4>Template Error</h4>%s</div>',
						$err_msg
					)
				);
						
		}

		// finally all checks are ok, return the template-array
		return $template;
	
	}
	
}

/**
 * 
 * The concrete class WP Simple HTML insert the data in the templates
 * @author Ralf Albert
 *
 */
class WP_Simple_HTML extends WP_Simple_Templater
{
	
	/**
	 * 
	 * Constructor overrides the delimiters in parent class
	 * @param WP_Simple_Templates $templates
	 */
	public function __construct( WP_Simple_Templates $templates ){
		
		parent::__construct( $templates );
		
		$this->set_delimiter( '{', '}' );
		
	}
	
	/**
	 * 
	 * Print the table userprofiles
	 * @param array $args
	 * @access public
	 * @since 0.1
	 */
	public function print_table_userprofile( $args ){

		echo $this->get_table_userprofile( $args );
		
	}
	
	/**
	 * 
	 * Create the table userprofile with the given arguments
	 * @param array $args
	 * @access public
	 * @since 0.1
	 */
	public function get_table_userprofile( $args ){
		
		if( ! is_array( $args ) )
			$args = (array) $args;

		$template = $this->get_templates( 'table_userprofile' );
		
		if( is_wp_error( $template ) )
			return $template->get_error_message();
				
		
		$defaults = array(
		
			'headline'		=> 'Headline',
		
			'host_id'		=> 'host_id',
			'host_label'	=> 'Label For Host',
			'host_value'	=> 'localhost:21',
			'host_desc'		=> 'Description for hostname',
		
			'uname_id'		=> 'uname_id',
			'uname_label'	=> 'Label For Uname',
			'uname_value'	=> 'UserName',
			'uname_desc'	=> 'The description for username',
		
			'ftppass_id'	=> 'ftppass_id',
			'ftppass_label'	=> 'Label For FTPPass',
			'ftppass_value'	=> 'Your ftp-password',
			'ftppass_desc'	=> 'The description for ftp-password',
		
		);
		
		$args = array_merge( $defaults, $args );
		
		return self::sprintf( $template['table'], $args );
		
	}

}


/* ------------------------------------------------------------------------------ */
/* ---------- Templates -------------------------------------------------- */
/* ------------------------------------------------------------------------------ */

/**
 * 
 * Abstract class WP Simple Templates define a method to retrive the defined templates
 * and a method to return all available templates
 * @author Ralf Albert
 *
 */
abstract class WP_Simple_Templates
{
	
	/**
	 * 
	 * Returns the template if it was defined in the template class. If the requested template was not
	 * defined in the template-class, return a wp-error.
	 * @param string $type
	 * @return array|string|object Array with template-strings or a single template-string or wp-error on failure
	 * @access public
	 * @since 0.1.2
	 */
	public function get_templates( $type = '' ){
		
		if( '' == $type )
			return NULL;
		
		// if the template-class have a method $type, return the template(s). else return a wp-error
		if( method_exists( $this, $type . '_template' ) ){
			
			$method = $type . '_template';
			return $this->$method();
			
		} else {
		
			return new WP_error(
				'template_error',
				printf(
					'Template <strong>%s</strong> does not exists in <strong>%s</strong>',
					$type,
					get_class( $this )
				)
			);

		}
			
	}
	
	/**
	 * 
	 * Return an array with all available templates from the template-class
	 * @return array $templates Array with available templates
	 * @access public
	 * @since 0.1.2
	 */
	public function get_available_templates(){

		// get all methods from the abstract class (__CLASS__) and the extended class ($this)
		$self_methods		= get_class_methods( __CLASS__ );
		$extended_methods	= get_class_methods( $this );
		
		// remove the '_template' extension and return the array with template-names
		$templates = array();
		
		foreach( array_diff( $extended_methods, $self_methods ) as $template )
			array_push( $templates, str_replace( '_template', '', $template ) );
			
		return $templates;
		
	}
	
}

/**
 * 
 * Concrete class WP Simple HTML Templates defines the templates
 * @author Ralf Albert
 *
 */
class WP_Simple_HTML_Templates extends WP_Simple_Templates
{
	
	/**
	 * 
	 * The template for table userprofile
	 */
	public function table_userprofile_template(){

		return '<h3>{headline}</h3>
				
				<table class="form-table">
					<tr>
						<th><label for="{host_id}">{host_label}</label></th>
						
						<td>
							<input type="text" name="{host_id}" id="{host_id}" value="{host_value}" class="regular-text" /><br />
							<span class="description">{host_desc}</span>
						</td>
						
					</tr>
				
					<tr>
						<th><label for="{uname_id}">{uname_label}</label></th>
						
						<td>
							<input type="text" name="{uname_id}" id="{uname_id}" value="{uname_value}" class="regular-text" /><br />
							<span class="description">{uname_desc}</span>
						</td>
						
					</tr>

					<tr>
						<th><label for="{ftppass_id}">{ftppass_label}</label></th>
						
						<td>
							<input type="password" name="{ftppass_id}" id="{ftppass_id}" value="{ftppass_value}" class="regular-text" /><br />
							<span class="description">{ftppass_desc}</span>
						</td>
					
					</tr>
				</table>';

	}
		
}

index.php
<?php
+/**
 * WordPress Plugin to store ftp-account data in the WordPress database
 *
 * PHP version 5.3
 *
 * @category   PHP
 * @package    WordPress
 * @subpackage FTP-Account to UserMeta
 * @author     Ralf Albert <me@neun12.de>
 * @license    GPLv3 http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.1.2
 * @link       http://wordpress.com
 */

/**
 * Plugin Name:	FTP-Account to UserMeta
 * Plugin URI:	https://github.com/RalfAlbert/WordPress-FTP-Account-to-UserMeta
 * Description:	This plugin add a couple of fields to your user-profile to enter your ftp-account data. This will prevent to re-entering your ftp-account data everytime WordPress request them.
 * Version: 	0.1.2
 * Author: 		Ralf Albert
 * Author URI: 	http://yoda.neun12.de
 * Text Domain: ftp2um
 * Domain Path: /languages
 * Network:
 * License:		GPLv3
 */

/*  Copyright 2012 Author  (email : me@neun12.de)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 3, as 
    published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


! defined( 'ABSPATH' ) and die( "Cheetin' uh!?" );

add_action( 'plugins_loaded', array( 'WP_FTPAcc_to_UM', 'plugin_start' ) );

register_activation_hook(	__FILE__, array( 'WP_FTPAcc_to_UM', 'activate_plugin' ) );
register_deactivation_hook(	__FILE__, array( 'WP_FTPAcc_to_UM', 'deactivate_plugin' ) );
register_uninstall_hook(	__FILE__, array( 'WP_FTPAcc_to_UM', 'uninstall_plugin' ) );

if( ! class_exists( 'WP_FTPAcc_to_UM' ) ){
	
	class WP_FTPAcc_to_UM
	{
		
		/**
		 * 
		 * The minimal role OR capability a user must have to enter the ftp-credentials
		 * 
		 * You can use a role or a capability to specify who can enter (and use) ftp-credentials.
		 * 
		 * Specifying a ROLE (e.g. administrator, contributor or reader) means,
		 * the user must have EXACT this role to use ftp-credentials. Use roles
		 * if you want to pick a small group of users.
		 * 
		 * Specifying a CAPABILITY (e.g. update_core, publish_posts or read) means,
		 * the user must have AT LEAST this capability to use ftp-credentials. Use capability
		 * if you want a larger group of users. 
		 * 
		 * @var string
		 */
		const MIN_LEVEL = 'administrator';

		/**
		 * 
		 * The metakey to save the ftp-credentials in usermeta
		 * @var string
		 */
		const METAKEY = 'FTP_Credentials';
		
		/**
		 * 
		 * Constant for textdomain
		 * @var string
		 */
		const LANG = 'ftp2um';
		
		/**
		 * 
		 * Container for the different template-engines
		 * @var object $template_engine
		 */
		protected static $template_engine = NULL;
		
		/**
		 * 
		 * ID for ftp-hostname in formular
		 * @var string $ftp_host
		 */
		protected $ftp_host = 'ftp_host';
		
		/**
		 * 
		 * ID for ftp-username in formular
		 * @var string $ftp_uname
		 */
		protected $ftp_uname = 'ftp_uname';
		
		/**
		 * 
		 * ID for ftp-password in formular
		 * @var string $ftp_pass
		 */
		protected $ftp_pass = 'ftp_pass'; 
		
		/**
		 * 
		 * Initialize a instance of the main-class
		 * @access public static
		 * @since 0.1
		 */
		public static function plugin_start(){
			/*
			 * In some circumstances, the global current user is not yet defined.
			 * If you experience problems with this, try declaring the global $current_user and
			 * using user_can($current_user->ID, $capability) - this seems to work more consistently. 
			 */
			
			
			// start the plugin only in backend and if the user have the minimum role/capability			
			if( is_admin() && current_user_can( self::MIN_LEVEL ) )
				new self();
				
			unset( $user );
				
		}

		/**
		 * 
		 * Things to do on plugin-activation
		 * @access public static
		 * @since 0.1
		 */
		public static function activate_plugin(){

			// need the autoloader
			self::init_autoloader();
			
			// check environment
			$v = new stdClass();
			$v->wp = '3.0';
			$v->php = '5.2';
			
			new WP_Environment_Check( $v );
			
		}
		
		/**
		 * 
		 * Things to do on plugin-deactivation
		 * @access public static
		 * @since 0.1
		 */
		public static function deactivate_plugin(){}
		
		/**
		 * 
		 * Things to do on plugin-uninstall
		 * @access public static
		 * @since 0.1
		 */
		public static function uninstall_plugin(){
			
			global $wpdb;
			
			// remove the entries in usermeta done by the plugin
			$wpdb->query( 
				$wpdb->prepare( 
					"DELETE FROM $wpdb->usermeta
					 WHERE meta_key = '%s'",
					self::METAKEY
				)
			);
			
		} 
		
		/**
		 * 
		 * Constructor
		 * Add hooks&filters
		 * @access public
		 * @since 0.1
		 */
		public function __construct(){
			
			// initialize the autoloader
			self::init_autoloader();
			
			// initialize the template-engine to avoid strict-errors
			self::$template_engine = new stdClass();
			
			// load the textdomain
			load_plugin_textdomain( 
				self::LANG, 
				FALSE, 
				dirname( plugin_basename( __FILE__ ) ) . '/languages/'
			);
						
			// hook all actions for backend
			add_action( 'admin_init', array( &$this, 'add_admin_hooks' ) );
			
		}
				
		/**
		 * 
		 * Initialize the autoloader
		 * @access protected
		 * @since 0.1
		 */
		protected static function init_autoloader( $config = array() ){
			
			// get the class if it was not already loaded
			if( ! class_exists( 'WP_Autoloader' ) )
				require_once dirname( __FILE__ ) . '/classes/class-wp_autoloader.php';
			
			// setup the default values
			$defaults = array(  
					'abspath'			=> __FILE__,
					'include_path' 		=> array( 'classes', ),
					'class_extension'	=> '.php',
					'class_prefix' 		=> array( 'class-', ),
			);
			
			$config = array_merge( $config, $defaults );
			
			new WP_Autoloader( $config );
			
		}
		
		/**
		 * 
		 * Caching for the template-classes
		 * Create a WP-Error if the template-engine could not be initialize
		 * @param string $template_engine The requested template-engine
		 * @return object The requested template-engine if available
		 * @access protected
		 * @since 0.1
		 */
		protected function get_template_engine( $template_engine ){
			
			$template_engine = strtolower( $template_engine );
			
			// available template-engines and the template-classes
			$available_template_engines = array(
			
				'html'	=> array( 'WP_Simple_HTML', 'WP_Simple_HTML_Templates' )
			
			);
			
			/*
			 * uncomment this line if plugins have to override the template engines
			 */
			// $available_template_engines = apply_filters( 'ftp2db_available_template_engines', $available_template_engines );
			
			// check if the requested template-engine is available. if not, try to create it
			if( ! isset( self::$template_engine->$template_engine ) ){

				if( ! isset( $available_template_engines[$template_engine] ) ){
					
					return new WP_Error( 'template_engine', 'Unknown template-engine <strong>' . $template_engine . '</strong>' );
					
				} else {
						
					$tmpl_ng	= &$available_template_engines[$template_engine];
					$class		= &$tmpl_ng[0];
					$template	= &$tmpl_ng[1];
					
					self::$template_engine->$template_engine = new $class( new $template );
					
				}
				
			}
			
			return self::$template_engine->$template_engine;

		} 
		
		/**
		 * 
		 * Adding the filters for the admin-backend
		 * @access public
		 * @since 0.1
		 * @internal hooked by 'admin_init' in ::__constructor()
		 */
		public function add_admin_hooks(){
			
			add_action( 'show_user_profile', array( &$this, 'add_custom_user_profile_fields' ) );
			add_action( 'edit_user_profile', array( &$this, 'add_custom_user_profile_fields' ) );
			
			add_action( 'personal_options_update', array( &$this, 'save_custom_user_profile_fields' ) );
			add_action( 'edit_user_profile_update', array( &$this, 'save_custom_user_profile_fields' ) );

			// add a filter to 'request_filesystem_credentials' so we can override the 
			// request-credentials-formular
			add_filter( 'request_filesystem_credentials', array( &$this, 'ftp_credentials' ), 1, 0 );
				
		}
		
		/**
		 * 
		 * Initialize and output the table for ftp-username and ftp-password
		 * @param object $user
		 * @return void
		 * @access public
		 * @since 0.1
		 * @internal hooked by 'show_user_profile' and 'edit_user_profile' in ::add_admin_hooks()
		 */
		public function add_custom_user_profile_fields( $user ){

			// get the ftp-credentials from database
			$creds = unserialize( get_user_meta( $user->ID, self::METAKEY, TRUE ) );
			
			// create the hostname for the formular
			// if no hostname (and no port) was found, blank hostname & port
			$hostname_attr = ! empty( $creds['hostname'] ) ? sprintf( '%s:%s', $creds['hostname'], $creds['port'] ) : '';
							
			// initialize the vars for the table
			$args = new stdClass();
			
			$args->headline			= __( 'FTP-Account', self::LANG );
			
			$args->host_id			= $this->ftp_host;
			$args->host_label		= __( 'FTP Host&Port', self::LANG );
			$args->host_value		= esc_attr( $hostname_attr );
			$args->host_desc		= __( 'Please enter your ftp hostname and port. E.g. <code>mydomain.com:21</code>', self::LANG );
			
			$args->uname_id			= $this->ftp_uname;
			$args->uname_label		= __( 'FTP Username', self::LANG );
			$args->uname_value		= esc_attr( $creds['username'] );
			$args->uname_desc		= __( 'Please enter your ftp username', self::LANG );
			
			$args->ftppass_id		= $this->ftp_pass;
			$args->ftppass_label	= __( 'FTP Password', self::LANG );
			$args->ftppass_value	= esc_attr( $creds['password'] );
			$args->ftppass_desc		= __( 'Please enter your ftp password', self::LANG );

			// using the template-engine to print the table
			$html = $this->get_template_engine( 'html' );
			
			if( is_wp_error( $html ) ){
				
				printf(
					'<div class="error"><h4>Template Error</h4>%s in line %s</div>',
					$html->get_error_message(),
					__LINE__
				);
				
			} else {
				
				$html->print_table_userprofile( $args );
				
			}

		}

		/**
		 * 
		 * Save the data in the usermeta
		 * @param integer $user_id
		 * @access public
		 * @since 0.1
		 */
		public function save_custom_user_profile_fields( $user_id ) {
			
			if( ! current_user_can( 'edit_user', $user_id ) )
				return FALSE;
			
			// retrieving the data from the POST array and filter them
			$user = filter_input( INPUT_POST, $this->ftp_uname, FILTER_SANITIZE_STRIPPED ); // only strip tags
			$pass = filter_input( INPUT_POST, $this->ftp_pass, FILTER_UNSAFE_RAW, FILTER_FLAG_STRIP_LOW ); // strip all characters below 32
			$host = filter_input( INPUT_POST, $this->ftp_host, FILTER_SANITIZE_STRIPPED ); // only strip tags

			// split host into hostname and port. if no port was given, set port to standard-port 21
			$host		= parse_url( $host );
			$hostname	= isset( $host['host'] ) ? $host['host'] : 
							( isset( $host['path'] ) ? $host['path'] : $host['scheme'] );
			$port		= isset( $host['port'] ) ? $host['port'] : '21';

			// save the ftp-credentials to database (usermeta). serialize the array to store a single value
			update_user_meta(
				$user_id,
				self::METAKEY,
				serialize(
					array(
						'hostname'	=> $hostname,
						'port'		=> $port,
						'username'	=> $user,
						'password'	=> $pass
					)
				)
			);
			
		}
		
		/**
		 * 
		 * Retrieve the ftp-credentials from the usermeta and serve them
		 * to the WordPress function "request_filesystem_credentials"
		 * If the user does not saved some ftp-account data, the formular to entering them will be displayed
		 * @access public
		 * @since 0.1
		 */
		public function ftp_credentials(){
			
			global $current_user;
			
			// try to get the credentials from the usermeta
			$creds_from_um = get_user_meta( $current_user->ID, self::METAKEY, TRUE );
			
			/*
			 *  check if any ftp-credentials are stored in usermeta. if not, return an empty string and display
			 *  the credentials-formular
			 *  
			 */
			if( empty( $creds_from_um ) )
				return '';
			else
				$creds = unserialize( $creds_from_um );

			$req_cred = array(
			
				'username'	=> '',
				'password'	=> '',
				'hostname'	=> '',
				'port'		=> '',
			
			);
			
			$req_cred = array_merge( $req_cred, $creds );
			
			/*
			 * if $req_cred contains some empty values, show the formular for ftp-credentials.
			 * else return the credentials from usermeta
			 * 
			 */
			foreach( $req_cred as $c )
				if( empty( $c ) )
					return '';
				
			return $req_cred;
			
		}
		
	} // .end class WP_FTPAcc_to_UM
	
	
} //.end if-class-exists WP_FTPAcc_to_UM